test/services/task_list_service_test.dart

```diff
@@ -1,47 +1,79 @@
 import 'package:flutter_test/flutter_test.dart';
 import 'package:p2p_task/models/task.dart';
-import 'package:p2p_task/services/identity_service.dart';
-import 'package:p2p_task/services/sync_service.dart';
-import 'package:p2p_task/services/task_list_service.dart';
-import 'package:p2p_task/utils/key_value_repository.dart';
-import 'package:sembast/sembast.dart';
-import 'package:sembast/sembast_memory.dart';
+import 'package:p2p_task/models/task_list.dart';
+import 'package:uuid/uuid.dart';
+
+import '../utils/device_task_list.dart';
 
 void main() {
-  late Database database;
-  late KeyValueRepository keyValueRepository;
-  late IdentityService identityService;
-  late SyncService syncService;
-  late TaskListService taskListService;
+  late DeviceTaskList device;
 
   setUp(() async {
-    database = await databaseFactoryMemory.openDatabase('');
-    keyValueRepository = KeyValueRepository(database);
-    identityService = IdentityService(keyValueRepository);
-    syncService = SyncService(keyValueRepository);
-    taskListService =
-        TaskListService(keyValueRepository, identityService, syncService);
+    device = await DeviceTaskList.create(name: 'device');
+  });
+
+  test('create task list', () async {
+    final id = Uuid().v4();
+    await device.taskListService
+        .upsertTaskList(TaskList(id: id, title: 'list1'));
+    final taskListRecordMap = await device.taskListService.taskLists;
+    expect(taskListRecordMap.map((e) => e.id!), [id]);
+    expect(taskListRecordMap.first.title, 'list1');
   });
 
   test('should store and retrieve tasks after first task is deleted', () async {
-    final taskTitle = 'Drink a cold cat';
-
-    await taskListService.upsert(
-      Task(
-        title: 'Catch a cat falling from the sky',
-        taskListID: '1',
-      ),
-    );
-    final task = (await taskListService.tasks).first;
-    await taskListService.remove(task);
-    await taskListService.upsert(Task(title: taskTitle, taskListID: '1'));
-    final tasks = await taskListService.tasks;
-
-    expect(tasks.length, equals(1));
-    expect(tasks.first.title, equals(taskTitle));
+    const taskTitle1 = 'Catch a cat falling from the sky';
+    const taskTitle2 = 'Drink a cold cat';
+
+    await device.taskListService
+        .upsertTaskList(TaskList(id: 'id1', title: 'list1'));
+    await device.taskListService.upsertTask('id1', Task(title: taskTitle1));
+    var tasks = (await device.taskListService.allTasks).toList();
+    expect(tasks.length, 1);
+    expect(tasks.first.title, taskTitle1);
+    final taskId = tasks.first.id!;
+    expect(taskId, isNot(null));
+
+    await device.taskListService.removeTask('id1', taskId);
+    tasks = (await device.taskListService.allTasks).toList();
+    expect(tasks.isEmpty, true);
+
+    await device.taskListService.upsertTask('id1', Task(title: taskTitle2));
```

Comment:
Deleting a task only marks the task as deleted (the task becomes a tombstone). I want to test if it can get changed back to an existing task.

Reply:
Understood. I would split this test into several tests then (also considering my comment above).
So write one test for each:

- test task id is set after creation
- test task can be retrieved from the database
- test task can be deleted from the database (and is marked as tombstone)
- test deleted task is recovered by creating it again

To avoid copy-pasting the task creation, either create the tasks in a setup method or in a helper.
Also, you could consider grouping the tasks (https://pub.dev/packages/test#writing-tests)